typescript/alert-center/src/utilities.ts

```diff
@@ -0,0 +1,165 @@
+import { ArgOptional, Transaction, TransactionContext} from '@dbos-inc/dbos-sdk';
+import { Knex } from 'knex';
+
+type KnexTransactionContext = TransactionContext<Knex>;
+
+export enum AlertStatus {
+  ACTIVE   = 0,
+  ASSIGNED = 1,
+  RESOLVED = 2
+}
+
+export interface Employee {
+  employee_name: string;
+  alert_id: number | null;
+  expiration: Date | null;
+  timeLeft?: number;
+}
+
+export interface AlertEmployee {
+  alert_id: number;
+  alert_status: AlertStatus;
+  message: string;
+  employee_name: string | null;
+}
+
+export interface AlertWithMessage {
+  alert_id: number;
+  alert_status: AlertStatus;
+  last_update_time: Date;
+  message: string;
+}
+
+export class RespondUtilities {
+  @Transaction({readOnly: true})
+  static async popDashboard(ctx: KnexTransactionContext) {
+    const alerts = await ctx.client<AlertEmployee>('alert_employee').select().orderBy(['alert_id']);
+    const employees = await ctx.client<Employee>('employee').select().orderBy(['employee_name']);
+    for (const a of alerts) {
+      if (a.alert_status === AlertStatus.ACTIVE && a.employee_name) {
+        a.alert_status = AlertStatus.ASSIGNED;
+      }
+    }
+    for (const p of employees) {
+      if (p.expiration) {
+        p.timeLeft = Math.round((p.expiration.getTime() - new Date().getTime())/1000);
+      }
+    }
+    return {alerts, employees};
+  }  
+
+  @Transaction()
+  static async cleanStaff(ctx: KnexTransactionContext) {
+    await ctx.client<Employee>('employee').whereNull('alert_id').delete();
+  }
+
+  @Transaction()
+  static async cleanAlerts(ctx: KnexTransactionContext) {
+    await ctx.client<AlertEmployee>('alert_employee').where({alert_status: AlertStatus.RESOLVED}).delete();
+  }
+
+  @Transaction()
+  static async getMaxId(ctx: KnexTransactionContext) {
+    const result = await ctx.client<AlertEmployee>('alert_employee').max('alert_id', { as: 'mid' }).first();
+    if (result ) {
+      return result.mid;
+    }
+    return -1;
+  }
+
+  @Transaction()
+  static async addAlert(ctx: KnexTransactionContext, message: AlertWithMessage) {
+    await ctx.client<AlertEmployee>('alert_employee').insert({
+      alert_id: message.alert_id,
+      alert_status: message.alert_status,
+      message: message.message,
+      employee_name: null,
+    }).onConflict(['alert_id']).ignore();
+  }
+
+  @Transaction()
+  static async getUserAssignment(ctx: KnexTransactionContext, employee_name: string, expiration: Date, @ArgOptional more_time: boolean | undefined) {
+    let employees = await ctx.client<Employee>('employee').where({employee_name}).select();
+    let newAssignment = false;
+    if (!employees.length) {
+      await ctx.client<Employee>('employee').insert({employee_name, alert_id: null, expiration: null});
+      employees = await ctx.client<Employee>('employee').where({employee_name}).select();
```

Comment:
We can simply `return` here. Because we'll get assignments next time we enter this function.

Reply:
Not quite. The alert could exist before the employee. There's no constraint that employees must be created before alerts. So I think it's cleaner to check if there's an assignment even if the employee just arrived.

Reply:
You just "insert" this employee and the transaction hasn't been committed yet, so other transactions cannot even see this newly inserted employee, and this "select" will always return "alert_id: null".

Reply:
Yes... other transactions cannot see this employee. And this query returns a structure with "alert_id: null". But it's also true that this transaction sees them and can complete the assignment and return it, no?

This query returns something like:
```
Employees [{"employee_name":"WhateverTheNewNameIs","alert_id":null,"expiration":null}]
```

Then, if we find an alert that needs assignment, we fill the other fields in. Something like:
```
Employees [{"employee_name":"WhateverTheNewNameIs","alert_id":1,"expiration":123456}]
```
Then we return that structure.

If we did do an early return here, the assignment would have to wait for another call, right? It would still happen eventually but slightly suboptimal. Or am I missing something?

Reply:
Yeah, if we return early the user might get their assignment next time. A minor optimization here is to use a `returning` clause in the insert query to avoid a separate "select" query to retrieve the just inserted data.